File: Pathogen_analysis.R
# Pathogen analysis script
#Author: Olabiyi Aderemi Obayomi
#E-mail: obadbotanist@yahoo.com
#created: February 2018

#remember to set your working directory to the directory containing your results together and your mapping file
######################## set all necessary variables   ####################################

# Set the necessary variables
file_name <- "wash_bac_pathogen_analysis"
independent_Variables <- c('WaterType', 'Plastic', 'Treatment', 'IrrigationType') 
mapping_file <- 'fruit_wash_pathogens_mapping.txt'
# Set to TRUE or FALSE if plots should be made
# I recommend setting it to FALSE when you don't detect a lot of pathogens in your samples
# for example when analysing for protists in the water and bacteria in the fruit puree
# but set to TRUE when you are analysing for bacteria pathogens in soil
make_plots <- FALSE 
Xlab <- c('Water Type','Plastic cover','Treatment', 'Irrigation Type')

# Source the libraries and functions needed for the analysis
source('Pathogen_analysis_functions.R')


# Change to the directory with all the results together / combined
setwd('results_together/')

# Get the pathogen file names arranged from the lowest to the highest numerically
pathogen_files <- list.files() #get all the files in the current directory
pathogen_files<- gsub(pattern = "__", 
                      replacement = ".",
                      x = mixedsort(gsub(pattern = "\\.", replacement = "__", x = pathogen_files)))

# Get the Treatment or sample names
# A character vector of sample or treatment names to be analysed as they appear in your mapping file
Treatments <- c() #intialize an empty treament vector
# Get a list of split file names by substititing "final_" with "" and splitting the file name by "_" 
split_file_name <- strsplit(x = gsub(pattern = "final_", replacement = "", x = pathogen_files),
                           split = "_" )
# Loop over the split_file_name list of lists getting the first element of each list
#and setting the respective treatment value to the right index in the vector of
#Treatments
for (index in 1:length(pathogen_files)) {
  Treatments[index] = split_file_name[[index]][1]
}

# Create a list of lists that will contain the outputs for the analysis
Result_together <- vector(mode = "list",length = length(Treatments))

#####################################################################################

### Perform pathogen analysis for each treatment or sample
# Turn off annoying warnings that could be ignored
options(warn = -1)
# A loop to perform the analysis for the number of treatments required
for (index in 1:length(Treatments)) {
  print(pathogen_files[index])
  Result_together[[index]] <- pathogen_analysis(treatment = Treatments[index], 
                                                pathogen_file = pathogen_files[index])
}
names(Result_together) <- Treatments
# Change to the parent directory
setwd('../')
# Turn warnings back on
options(warn = 0)

########## Get the frequency tables for all the taxon levels together in a list called all_frequency_tables #

# Taxon levels vector
taxons <- c('genera','species', 'strains')
all_frequency_tables <- vector(mode = "list",length = length(taxons))
# Combines the frequency tables for all the taxon levels
for (index in 1:length(taxons)) {
  
  taxon_frequency_table <- combine_freq_tables(Treatments = Treatments,
                                               Result_together = Result_together,
                                               frequency_table_name = taxons[index])
  all_frequency_tables[[index]] <- taxon_frequency_table
}
names(all_frequency_tables) <- taxons

##################################make taxon tables #####################
# Make raw and normalized taxon/otu tables for all the taxon levels and combine them
combined_taxon_tables <- make_taxon_tables(samples = Treatments, 
                                           taxons = taxons, 
                                           all_frequency_tables = all_frequency_tables)


################################ Get the number of pathogens detected per treament group #######
# Get the independent variables dataframe
independent_variables.df <- get_independent_variables_df(mapping_file = mapping_file, 
                                                         independent_variables = independent_Variables)

# Intialize a list to store the results of the pathogens detected per treatment group
detect <- vector(mode = 'list', length = length(independent_Variables))
# Get the taxon table
taxon_table <- combined_taxon_tables$raw_tables$species
# Get the frequency of detection for each pathogen per treatment group
for (index in 1:length(independent_Variables)) {
  
detect[[index]]<- presence_absence(taxon_table = taxon_table, 
                          independent_variables.df = independent_variables.df, 
                          variable = independent_Variables[index])
}
names(detect) <- independent_Variables

#write the results to an excel file per independent variable
for (i in names(detect)){
write.xlsx(x = detect[[i]], 
           file = sprintf(fmt = '%s.xlsx', file_name), 
           sheetName = i, 
           col.names = T, 
           row.names = F,
           append = TRUE)
}





if (make_plots){
  

############################### Plotting ##############################################
# Open a pdf device and set its width and height
pdf(file =sprintf('%s.pdf',file_name),width = 14, height = 10.5)

# Generate a list of lists that will contain ordination tables for each taxon level
ord_tables <- vector(mode = "list",length = length(taxons))
stats_table <- vector(mode = "list",length = length(taxons))
stacked_stats_table <- vector(mode = "list",length = length(taxons))
abund_table_norm <- vector(mode = "list",length = length(taxons))
# Make plots for each taxon level
for (taxon in 1:length(taxons)){
  # Make ordanition plots - PCOA and NMDS and return dataframes with the principal components
  ord_tables[[taxon]] <-ordination_plots(taxon_table = combined_taxon_tables[["normalized_tables"]][[taxon]],
                                         mapping_file = mapping_file, 
                                         independent_variables = independent_Variables,
                                         label =  'samples')
  
  # Get the relative abundance table per sample with the independent variables attached
  stats_table[[taxon]] <- ord_tables[[taxon]][[4]]
  num_of_numericVariables <- ncol(stats_table[[taxon]]) - length(independent_Variables)
  col_num_of_numericVariables <- (length(independent_Variables)+1):ncol(stats_table[[taxon]])
  stacked_stats_table[[taxon]]<- Stack_data(Data = stats_table[[taxon]], 
                                            lab_columns = independent_Variables,
                                            num_of_numericVariables = num_of_numericVariables , 
                                            col_num_of_numericVariables = col_num_of_numericVariables)
  
  for (i in 1:length(independent_Variables)){ 
    # Make relative abundance boxplots faceted by pathogens with each sample info
    gg <- draw_pathogen_plots(Data = stacked_stats_table[[taxon]], 
                              main = 'Difference between Pathogens',
                              x = independent_Variables[i], 
                              y = 'Relative.abundance', 
                              Xlab = Xlab[i], 
                              YLab = 'Relative abundance', 
                              plotType = 'box', 
                              group = 'Pathogens')
    invisible(print(ggpar(gg, tickslab = FALSE, ticks = FALSE, xlab = FALSE)))
    # Make long relative abundance box plots
    long_gg<- ggboxplot(data = stacked_stats_table[[taxon]], x = 'Pathogens', 
                        y = 'Relative.abundance', 
                        #fill = independent_Variables[i],
                        color = independent_Variables[i],
                        add = 'mean_se')
    invisible(print(ggpar(long_gg,x.text.angle = 90, 
          rotate = TRUE, 
          font.x = c(14, "bold", "black"),
          font.y = c(14, "bold", "black"),xlab = taxons[taxon], ylab = 'Relative abundance')))
    
    # Generate stacked frequency and abundance tables of an independent variable 
    # first check if the independent variable is not a dataframe
    independent_variables<- stats_table[[1]][,1:length(independent_Variables)]
    if (is.data.frame(independent_variables) == FALSE){
      independent_variables<- as.data.frame(independent_variables)
      rownames(independent_variables) <- rownames(stats_table[[1]])
      colnames(independent_variables) <- independent_Variables
    }
    # Generate stacked frequency and abundance tables of an independent variable
    abund_table_norm[[taxon]][[i]] <- generate_abundance_tables(independent_variables = independent_variables, 
                                                                variable = independent_Variables[i], 
                                                                taxon_table = combined_taxon_tables[["raw_tables"]][[taxon]])$relative.abundance
    ## Make relative abunbance bar charts
    draw_pathogen_plots(Data = abund_table_norm[[taxon]][[i]], 
                        main = 'Pathogens relative abundance', 
                        x = independent_Variables[i], 
                        y = 'Relative.abundance', 
                        Xlab = Xlab[i], 
                        YLab = 'Relative abundance', 
                        plotType = 'bar', 
                        group = 'Pathogens')
    
  }
  
}
# Naming the lists by taxon levels
names(ord_tables) <- taxons
names(stats_table) <- taxons
names(stacked_stats_table) <- taxons
names(abund_table_norm) <- taxons




########################## Make pathogen Heat map #########################################
# Get the raw table
x <- generate_abundance_tables1(independent_Variables = independent_Variables,
                                variable = independent_Variables[1],
                                taxon_table = combined_taxon_tables[["raw_tables"]][[2]],
                                stats_table = species_stats_table)$raw_table

# Rename the rownames of the taxon tables 
rownames(x) <- x[,1]
# Remove the redundant treatment column
x <- x[,-1]
# Normalize the taxon table
x_norm <- normalize_otu_table(OTU_table = x, method = "logUQ")
# Convert relative abundance
x_norm.abund <- as.matrix(sweep(x = x, MARGIN = 1, rowSums(x_norm),'/'))
x_norm.abund <- t(x_norm.abund)
# Get the indices of the ordered pathogen names
id<- order(rownames(x_norm.abund))
# Get their names based on the indices
id<- rownames(x_norm.abund)[id]
# Arrange the names alphabetically
x_norm.abund <- x_norm.abund[id, ]

# Reorder the columns of a matrix
# A vector to re-order the columns of the matrix
col.order <- c("PW", "TWW","non-irrigated", "PW-P", "PW+P", "TWW-P", "TWW+P", "Blank",
               "PW+SD-P", "PW+SD+P", "TWW+SD-P", "TWW+SD+P", "TWW+SSD-P", "TWW+SSD+P")
ordered_x_norm.abund <- x_norm.abund[,col.order]



# Make heatmap of normalized taxon table clustered using hclust on bray curtis distances
heatMap <- Heatmap(ordered_x_norm.abund, 
        cluster_columns = FALSE, 
        cluster_rows = FALSE, 
        column_title = 'Treatment', 
        row_names_side = 'left', 
        column_title_side = 'bottom', 
        row_names_gp = gpar(fontface="bold.italic",cex=0.6), 
        column_names_gp = gpar(fontface="bold"), 
        heatmap_legend_param = list(title= 'Relative abundance'))





############################################################################################
}
#turn-off all open devices
dev.off()

# String vector specifying the sheet name in an excel file
Matrix <- 'melon'
# Get the species taxon table
species_otu_table <- combined_taxon_tables[[1]][[2]]
# Write the species taxon table to an excel file
write.xlsx(x = t(species_otu_table), file = '../species_table.xlsx', sheetName = Matrix, col.names = T, row.names = T, append = T)

File: Pathogen_analysis.py
# -*- coding: utf-8 -*-
"""
Created on Sat Nov 24 21:24:15 2018

@author: Olabiyi Aderemi Obayomi
@email: obadbotanist@yahoo.com


"""

import os
import pandas as pd
import seaborn as sns
from skbio.stats.ordination import pcoa
import pathogenAnalysisFunctions as p
from plotnine import *
import warnings
import random


########### Set the required input variables (an example) 
file_name = "barrier_bacteria" 
MAP = "barrier_bacteria_mapping.txt"
independent_variables = ["WaterType","Plastic","Treatment","Season"]
sample_column = "sample" # your mapping file must have a column called sample
working_directory = (r"C:/Users/user/Documents/programming/Python/pathogen_analysis/barrier/results_together/")
group="Treatment"
# PCOA shape and color labels
color_var='WaterType'
shape_var='Plastic'
color_lab= 'Water Type',
shape_lab = 'Plastic'
point_colors=['blue','purple']
# Set color map for the heatmaps
cmap= 'hot' #["white","pink","red"] #


# Define a publication ready theme for making ggplots
publication_format = (theme_bw() +
  theme(panel_grid = element_blank()) +
  theme(axis_ticks_direction='in',
  axis_ticks_length=3.5,
  axis_text_x=element_text(margin={'t':0.1,'r':0,'b':0,'l':0,'units':"in"}),
  axis_text_y=element_text(margin={'t':0,'r':0.1,'b':0,'l':0,'units':"in"}), 
  axis_title = element_text(size = 18, style='italic', weight='bold', color = 'black'), 
  axis_text = element_text(size = 16, style ='italic', weight='bold', color = 'black'),
  legend_position = 'right', 
  legend_text = element_text(size = 14, weight ='bold', color = 'black'),
  legend_title = element_text(size = 15, weight ='bold', color = 'black'),
  strip_text =  element_text(size = 14, weight ='bold', color = 'black')))

# Generate random colors and add them to the custom palette below when you
#don't have enough colors for making your stack bar plot
random.seed(1)
#generate more colors and test if it works 
_,more_colors= p.generate_colors(n= 8)

custom_palette = ["#a6cee3","#1f78b4","#b2df8A","#33a02c","#fb9a99","#e31a1c","#fdbf6f", "#ff7f00",
                  "#6A3D9A","#FF00FFFF","#B15928","#000000","#FFC0CBFF","#8B864EFF","#F0027F",
                  "#666666","#1B9E77", "#E6AB02","#A6761D","#FFFF00FF","#FFFF99","#00FFFFFF",
                  "#B2182B","#FDDBC7","#D1E5F0","#CC0033","#FF00CC","#330033","#999933","#FF9933",
                  "#FFFAFAFF",'#988ea3','#b8aec3','#d8cee3','#382e43','#584e63','#786e83','#5726ac',
                  '#7746cc','#9766ec','#d7a62c', '#f7c64c','#17e66c','#49b926','#69d946','#89f966',
                  '#a91986','#c939a6','#e959c6', '#42f8e3', '#62183','#823823', '#a25843', '#c27863',
                  '#e29883', '#2b8a3', '#22d8c3']





# Turn off annoying warnings
warnings.filterwarnings('ignore')

######## Run pathogen analysis and combine the results for all samples
# then return the combined results and the sample names
os.chdir(working_directory)
files = os.listdir()
result_together,samples = p.combined_results(files)

# Make pathogen taxon table 
taxon_table = p.make_taxon_table(result_together, samples)

# Change to the parent directory
os.chdir("../")

# Write-out the pathogen table to excel 
writer = pd.ExcelWriter("{}_table.xlsx".format(file_name))
taxon_table.to_excel(writer,sheet_name="taxon_table") 


# Get the independent variables dataframe
mapping_file = pd.read_csv(MAP, sep="\t") 
independent_variables.append(sample_column)
independent_variable_df = mapping_file[independent_variables].set_index(sample_column)
independent_variables.remove(sample_column)

# Get the transposed pathogen dataframe and independent variables with common ids
# by joining the independent variable dataframe with the pathogen dataframe
# based on samples common to both dataframes
taxon_table_t = taxon_table.T

independent_variable_df, taxon_table_t, abund_df = p.join_tables(table1=independent_variable_df,
                                                               table2=taxon_table_t)

# Get the number of pathogens detected per treatment and 
# write the resulting table to excel
detect_df = p.presence_abscence(taxon_table_t=taxon_table_t,
                                independent_variables=independent_variable_df)
detect_df.to_excel(writer,sheet_name="detects")

# Remove samples that pathogens were no detected in
taxon_table_t = p.remove_zero_samples(taxon_table_t=taxon_table_t)
# Estimate diversity and write the diversity table to excel
diversity_table = p.estimate_diversity(taxon_table_t= taxon_table_t)
diversity_table.to_excel(writer,sheet_name="diversity")

# Save the excel writer object for the open excel file
writer.save()


# Convert the absolute pathogen taxon table to a relative abundance table
abund_table = taxon_table_t.apply(func=lambda x: x/sum(x),axis=1)
# Drop samples with NA having zero counts
abund_table = abund_table.dropna()



# Calculate bray curtis dissimilarity matrix
brayMat = p.table_to_distances(abund_table.T,p.bray_curtis_distance)
# Perform principal co-ordinate analysis
bc_pc= pcoa(brayMat)

# Extract the first two principle components and combine
# them with the independent variables dataframe
pc_df=bc_pc.samples[["PC1","PC2"]]

independent_variable_df,_,pc_df = p.join_tables(table1=independent_variable_df,
                                                               table2=pc_df)

proportion_explained = bc_pc.proportion_explained[0:2] * 100

xlab = "PCO1({:.2f}%)".format(proportion_explained["PC1"])
ylab = "PCO2({:.2f}%)".format(proportion_explained["PC2"])

# Make the PCoA plot and save it as a png file 
gg = (ggplot(data=pc_df) + 
 geom_point(aes(x='PC1', y='PC2', color=color_var, shape=shape_var), size=6) + 
 labs(x=xlab,y=ylab,color= color_lab, shape= shape_lab) + 
 scale_color_manual(values=point_colors) + publication_format)
gg

ggsave(plot=gg, filename='{}_pcoa.png'.format(file_name),
       width=14, height=8, dpi=600, device='png')



## Prepare data for plotting relative abundances based on treatment category ##########

# Aggregate by group and then stack the data
treat_abund_table = p.get_treatment_abund_table(independent_variable_df, taxon_table_t, group=group)


treat_rare_abund_table=p.remove_low_abundance_pathogens(abun_table=treat_abund_table,
                                               group_low_abund=True,
                                               threshold=0.02)
treat_rare_abund_table[group] = treat_rare_abund_table.index

stacked_treat_abund_table = treat_rare_abund_table.melt(id_vars=[group], 
                                           value_vars=list(treat_rare_abund_table.columns.drop(group)), 
                                           value_name="Relative_abundance", 
                                           var_name="Pathogens")


# Make stacked barplot

gg = (ggplot(data= stacked_treat_abund_table) +
geom_col(aes(x=group,y="Relative_abundance", fill="Pathogens")) + 
 labs(y="Relative abundance(%)") +
 scale_fill_manual(values=custom_palette) +
 publication_format)
gg
ggsave(plot=gg, filename='{}_bar.png'.format(file_name),
       width=14, height=8, dpi=600, device='png')


# Make box plots

gg =(ggplot(data= stacked_treat_abund_table) + 
    geom_boxplot(aes(x="Pathogens",y="Relative_abundance", color=group)) +
    labs(y="Relative abundance(%)") +
    publication_format + theme(axis_text_x=element_text(angle=45)))
gg
ggsave(plot=gg, filename='{}_boxplot.png'.format(file_name),
       width=14, height=8, dpi=600, device='png')

# Prepare to make heatmaps
# Optionally remove pathogens that are not atleast 5% abundant in a sample
abund_pathogens_table = p.remove_low_abundance_pathogens(
                                              abun_table=treat_abund_table,
                                              group_low_abund=False)

# Sort the columns/pathogens in alphabetical order
col_names = abund_pathogens_table.columns.tolist()
col_names.sort()
# Re-arrange the columns/pathogens in alphabetical order
abund_pathogens_table = abund_pathogens_table[col_names]

##### Make heatmaps

# Heatmap for all the pathogens
col_names = treat_abund_table.columns.tolist()
# Sort the columns/pathogens in alphabetical order
col_names.sort()
# Re-arrange the columns/pathogens in alphabetical order
treat_abund_table = treat_abund_table[col_names]
fig,r = plt.subplots(figsize=(14,7))
sns.heatmap(treat_abund_table.T, cmap=cmap,\
              linecolor="black", linewidths=0.01, yticklabels=True, ax=r)
plt.show(r)
# Save the heatmap to a png file
r.figure.savefig("all_pathogens_heatmap.png", dpi=600)

# Heatmap for only the abundant pathogens
fig,a = plt.subplots(figsize=(14,7))
sns.heatmap(abund_pathogens_table.T, cmap=cmap,\
              linecolor="black", linewidths=0.01,
            ax=a, yticklabels=True)
plt.show(a)
# Save the heatmap to a png file
a.figure.savefig("abundant_pathogens_heatmap.png", dpi=600)

# Restore warnings
warnings.filterwarnings('default')